### ISSUE TITLE
[Bug]: Memory leak observed after merging 0731 release with latest patches

### ISSUE THREAD

**Issue by snarunner** [yhfgyyf/vllm-deepseek-v4-sm89#49]
## Issue Description
A system memory (Host RAM) leak occurs when running the service using the 0731 version merged with the latest upstream patches/commits. 

Once the service starts running, system memory usage grows continuously over time without stabilization, eventually leading to OOM / system unresponsiveness.

---

## Environment Info
- **Base Version:** 0731 release / tag
- **Patch Status:** Cherry-picked / merged latest upstream patches
- **OS:** Linux (Ubuntu / RHEL)
- **Deployment:** Systemd / Docker container
- **Hardware:** CPU / GPU setup

---

## Expected Behavior
System memory (RAM) usage should remain stable or bounded after warming up during normal runtime.

## Actual Behavior
System memory continuously increases linearly/monotonically during operation until system RAM is exhausted.

---

## Steps to Reproduce
1. Build / deploy using the `0731` base code merged with the latest patches.
2. Launch the runtime process / engine.
3. Send continuous workload / requests or let it run idling under load.
4. Monitor host memory (e.g., via `htop` / `free -m` / `ps aux`). System memory rises continuously without being released.

---

## Additional Context / Logs
- Memory growth rate: *(e.g., grows ~X GB every hour)*
- GPU VRAM status: *(e.g., VRAM usage remains stable; leak is strictly isolated to Host RAM / C++ heap / Python process memory)*
- *(Optional) Any heap profile or `valgrind` / `gperftools` / `tracemalloc` logs if available.*

## Issue Description
A system memory (Host RAM) leak occurs when running the service using the 0731 version merged with the latest upstream patches/commits. 

Once the service starts running, system memory usage grows continuously over time without stabilization, eventually leading to OOM / system unresponsiveness.

---

## Environment Info
- **Base Version:** 0731 release / tag
- **Patch Status:** Cherry-picked / merged latest upstream patches
- **OS:** Linux (Ubuntu / RHEL)
- **Deployment:** Systemd / Docker container
- **Hardware:** CPU / GPU setup

---

## Expected Behavior
System memory (RAM) usage should remain stable or bounded after warming up during normal runtime.

## Actual Behavior
System memory continuously increases linearly/monotonically during operation until system RAM is exhausted.

---

## Steps to Reproduce
1. Build / deploy using the `0731` base code merged with the latest patches.
2. Launch the runtime process / engine.
3. Send continuous workload / requests or let it run idling under load.
4. Monitor host memory (e.g., via `htop` / `free -m` / `ps aux`). System memory rises continuously without being released.

---

## Additional Context / Logs
- Memory growth rate: *(e.g., grows ~X GB every hour)*
- GPU VRAM status: *(e.g., VRAM usage remains stable; leak is strictly isolated to Host RAM / C++ heap / Python process memory)*
- *(Optional) Any heap profile or `valgrind` / `gperftools` / `tracemalloc` logs if available.*

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by yhfgyyf** [comment]
PR #51 fixes the paged MQA illegal-memory-access issue, but it does not address the host RAM leak reported here. Please provide a minimal reproduction and memory profile.

**Comment by snarunner** [comment]
Thanks for the clarification. The issue is reproducible with a combined patch set: I applied the repository's latest patches on top of the int64 patch. The running image is `local/vllm-dsv4-sm89:v0.23.1rc1.dev904-g8e321cc4f-sm89.1-int64`; this is not a test of PR #51 alone.

## Host / runtime

- OS: Ubuntu 22.04.5 LTS, kernel 6.8.0-110-generic
- CPU: 2 × Intel Xeon Gold 6454S (128 logical CPUs total)
- RAM: 251 GiB; swap disabled
- GPU: 8 × NVIDIA GeForce RTX 4090 D, 49,140 MiB each
- NVIDIA driver: 595.58.03
- Deployment: Docker with `--gpus all` and `--ipc=host`

## Minimal reproduction

I start one 8-way tensor-parallel server with the combined image and then send continuous traffic after warm-up:

```bash
docker run --gpus all --ipc=host \
  -e CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
  local/vllm-dsv4-sm89:v0.23.1rc1.dev904-g8e321cc4f-sm89.1-int64 \
  /model --tensor-parallel-size 8 \
  --gpu-memory-utilization 0.90 \
  --max-num-batched-tokens 8192 --max-num-seqs 32 \
  --max-model-len 204800 --kv-cache-dtype fp8 --block-size 256 \
  --no-enable-flashinfer-autotune \
  --enable-prefix-caching --enable-chunked-prefill \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 --reasoning-parser deepseek_v4 \
  --attention-backend FLASHINFER_MLA_SPARSE_DSV4
```

The observed symptom is host-RAM growth over time under continuous requests, while GPU allocation does not show the same unbounded pattern.

For a current runtime snapshot, the engine process RSS is about 1.2 GiB and each of the eight worker processes is about 4.5 GiB RSS. GPU memory at the snapshot is approximately 44.26 GiB / 47.99 GiB per GPU. Host memory is 41 GiB used out of 251 GiB at that point. I can provide a timestamped RSS/host-memory time series if that would help isolate the growth.

**Comment by snarunner** [comment]
## Update: root cause, permanent fix, and complete image patch inventory

I isolated this issue to **Triton JIT specialization on request-varying shapes/strides**, rather than PR #51 or the paged-MQA illegal-memory-access fix.

Several DeepSeek-V4 Triton kernel arguments that change with request length were declared as <code>tl.constexpr</code>. Triton therefore retained a different compiled kernel/cubin for many distinct request shapes for the lifetime of the worker. Under variable-length production traffic this cache keeps growing in host RAM and can eventually cause OOM or a hard freeze.

The relevant upstream work is the still-open [vllm-project/vllm#41834](https://github.com/vllm-project/vllm/pull/41834). Its 2026-07-04 stable-preview notes explicitly describe the same per-shape recompilation/unified-memory leak.

### Required upstream patch set

The SM89 base used here needs the following **three commits together**:

1. [8571496a4ce53a32b6bb268e630b74f737327c7d](https://github.com/jasl/vllm/commit/8571496a4ce53a32b6bb268e630b74f737327c7d)  
   Makes o_proj FP8-einsum <code>num_tokens</code> and its dependent strides runtime values and adds <code>do_not_specialize</code>. This collapses distinct cubins across different prefill token counts to one kernel.

2. [15434cb64f5f95b9118d3dd1c2695885d2da2753](https://github.com/jasl/vllm/commit/15434cb64f5f95b9118d3dd1c2695885d2da2753)  
   Makes MQA-logits <code>num_q</code>, <code>seq_len_kv</code>, <code>num_rows</code>, <code>logits_width</code>, and <code>stride_lm</code> runtime values. It also restores the provable 16-byte alignment hints with <code>tl.multiple_of</code>, avoiding the reported decode-performance loss without putting request shapes back into the JIT key.

3. [b43470e871590a33cf5689a4470b75cc6792ce6a](https://github.com/jasl/vllm/commit/b43470e871590a33cf5689a4470b75cc6792ce6a)  
   Makes TF32 prenorm <code>stride_outs</code> and <code>stride_sqs</code> runtime values. This is a smaller contributor, mainly visible at very long context.

The complete upstream comparison is [444fe3ac8b...b43470e871](https://github.com/jasl/vllm/compare/444fe3ac8b...b43470e871).

This is an **SM89 backport/adaptation**, not a blind cherry-pick. The current SM89 base did not yet contain the prerequisite from commit 8571496, so applying only the last two commits would leave one leak path. Conversely, the SM89 fork contains an Ada-specific <code>UPCAST_FP8</code> path that must be preserved. The upstream fused-quant einsum hunk is not applied because this older base does not contain that kernel.

### Complete image-layer inventory

Original base, left unchanged:

- Image: <code>local/vllm-dsv4-sm89:v0.23.1rc1.dev904-g8e321cc4f-sm89.1-int64</code>
- Image ID: <code>sha256:12f16be0a50d85c6d27a693bd176582332f14d3af4b12d6ef093f29e78cbdedd</code>
- vLLM wheel: <code>0.23.1rc1.dev904+g8e321cc4f.cu130</code>
- FlashInfer wheel: <code>0.6.14+sm89.1</code>; <code>flashinfer-cubin==0.6.13</code>
- Existing <code>sm89-long-context-int64.patch</code>: inherited unchanged.

New derived image:

- Image: <code>local/vllm-dsv4-sm89:v0.23.1rc1.dev904-g8e321cc4f-sm89.1-int64-tritonleakfix1</code>
- Image ID: <code>sha256:8c8c0acb930ece8dfd68a0c36512bf9e752256ac85b81f9a92bc56aee6d73cb4</code>
- Built **directly from the original int64 image**, not from the previous workspace-fix image.
- Newly added functional changes: the adapted effects of the three commits above.
- Runtime source files changed: only
  - <code>vllm/models/deepseek_v4/nvidia/ops/fp8_einsum.py</code>
  - <code>vllm/models/deepseek_v4/nvidia/ops/sm12x_mqa.py</code>
- Preserved: the original INT64 long-context patch and SM89 <code>UPCAST_FP8</code> logic.
- Explicitly **not included**: <code>df71917cf17c97d81f48122e6baa7c13d184ff90</code> / <code>workspacefix1</code>. That patch concerns eager-break workspace reuse and did not stop this per-shape Triton cache growth.
- No service flags, model configuration, or launch configuration are part of this image patch.

For clarity, this is the complete runtime patch applied on top of the original int64 image:

~~~diff
diff --git a/vllm/models/deepseek_v4/nvidia/ops/fp8_einsum.py b/vllm/models/deepseek_v4/nvidia/ops/fp8_einsum.py
--- a/vllm/models/deepseek_v4/nvidia/ops/fp8_einsum.py
+++ b/vllm/models/deepseek_v4/nvidia/ops/fp8_einsum.py
@@ -13,23 +13,32 @@
-@triton.jit
+# Request token counts and their dependent strides must stay runtime values.
+# Otherwise Triton retains one compiled kernel for every request shape.
+@triton.jit(
+    do_not_specialize=[
+        "num_tokens",
+        "a_stride_group",
+        "a_scale_stride_group",
+        "a_scale_stride_hidden",
+    ]
+)
 def _deepseek_v4_sm12x_fp8_einsum_kernel(
     a_ptr,
     a_scale_ptr,
     b_ptr,
     b_scale_ptr,
     out_ptr,
-    num_tokens: tl.constexpr,
+    num_tokens,
     num_groups: tl.constexpr,
     out_rank: tl.constexpr,
     hidden_size: tl.constexpr,
     a_stride_token: tl.constexpr,
-    a_stride_group: tl.constexpr,
+    a_stride_group,
     a_stride_hidden: tl.constexpr,
     a_scale_stride_token: tl.constexpr,
-    a_scale_stride_group: tl.constexpr,
-    a_scale_stride_hidden: tl.constexpr,
+    a_scale_stride_group,
+    a_scale_stride_hidden,
     b_stride_group: tl.constexpr,
     b_stride_out: tl.constexpr,
     b_stride_hidden: tl.constexpr,
@@ -44,6 +53,11 @@
     BLOCK_HIDDEN: tl.constexpr,
     UPCAST_FP8: tl.constexpr,
 ) -> None:
+    # Preserve vectorized loads without putting request-dependent values
+    # back into the JIT specialization key.
+    a_stride_group = tl.multiple_of(a_stride_group, 16)
+    a_scale_stride_group = tl.multiple_of(a_scale_stride_group, 16)
+
     token_block = tl.program_id(0)

diff --git a/vllm/models/deepseek_v4/nvidia/ops/sm12x_mqa.py b/vllm/models/deepseek_v4/nvidia/ops/sm12x_mqa.py
--- a/vllm/models/deepseek_v4/nvidia/ops/sm12x_mqa.py
+++ b/vllm/models/deepseek_v4/nvidia/ops/sm12x_mqa.py
@@ -60,8 +60,8 @@
-    num_q: tl.constexpr,
-    seq_len_kv: tl.constexpr,
+    num_q,
+    seq_len_kv,
@@ -71,7 +71,7 @@
-    stride_lm: tl.constexpr,
+    stride_lm,
@@ -219,8 +219,8 @@
-    num_rows: tl.constexpr,
-    logits_width: tl.constexpr,
+    num_rows,
+    logits_width,
@@ -240,7 +240,7 @@
-    stride_lm: tl.constexpr,
+    stride_lm,
@@ -332,8 +332,8 @@
-    num_rows: tl.constexpr,
-    logits_width: tl.constexpr,
+    num_rows,
+    logits_width,
@@ -353,7 +353,7 @@
-    stride_lm: tl.constexpr,
+    stride_lm,
@@ -646,10 +646,10 @@
-    stride_outs: tl.constexpr,
+    stride_outs,
     stride_outm: tl.constexpr,
     stride_outn: tl.constexpr,
-    stride_sqs: tl.constexpr,
+    stride_sqs,
~~~

### Build and verification steps

1. Build a new layer with <code>FROM local/vllm-dsv4-sm89:v0.23.1rc1.dev904-g8e321cc4f-sm89.1-int64</code>; do not edit or retag the base image.
2. Apply the patch under <code>/opt/release-venv/lib/python3.12/site-packages</code> with <code>patch --batch --forward -p1</code>.
3. Run <code>py_compile</code> on both modified files.
4. Run AST assertions that all listed dynamic values no longer have <code>tl.constexpr</code>, that the four einsum arguments are in <code>do_not_specialize</code>, and that both <code>tl.multiple_of(..., 16)</code> hints remain.
5. Assert that <code>UPCAST_FP8: tl.constexpr</code> still exists and that the old <code>_dsv4_workspace_fix.so</code> artifact is absent.
6. Import both patched modules before completing the image build.
7. Record the three upstream SHAs, original base digest, and excluded workspace patch as OCI labels.

### Preliminary runtime verification

The derived image was then started with the existing production configuration:

- service reached ready state; <code>/health</code> returned HTTP 200;
- a real <code>/v1/chat/completions</code> request returned HTTP 200;
- logs confirmed the SM89 direct FP8 MQA path was active;
- the first live request produced one cold <code>_fp8_mqa_logits_kernel</code> JIT event, and subsequent different request shapes did not add more MQA JIT compilations in the observation window;
- over 20 samples / 9m33s with 24 additional completed requests, cgroup anonymous memory increased by about 4.3 MiB during warm-up and then stayed flat; total <code>memory.current</code> changed by about 5.8 MiB, <code>memory.peak</code> was unchanged, and there were no OOMs or worker restarts.

One monitoring detail matters: a large <code>memory.stat:file</code> value is Linux reclaimable model/checkpoint page cache, not process-private leaked memory. For this issue, use cgroup <code>anon</code>, process RSS/PSS, host <code>MemAvailable</code>, OOM counters, and Triton JIT event counts rather than treating total cgroup memory (including file cache) as the leak signal.

This short production sample confirms that the repeated MQA specialization stopped on the exercised path, but it is not yet a substitute for a multi-hour high-load soak across many prompt lengths. I will continue to treat the result as preliminary until that longer observation is complete.
